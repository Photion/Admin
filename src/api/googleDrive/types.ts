/**
 * Lists Google Drive OAuth2 scopes
 * used by Photion
 */
export enum DriveScopes {
  DRIVE__APPDATA = 'https://www.googleapis.com/auth/drive.appdata',
  DRIVE__FILE = 'https://www.googleapis.com/auth/drive.file',
  DRIVE__METADATA = 'https://www.googleapis.com/auth/drive.metadata',
}

/**
 * Models a File resource
 * from Google Drive
 */
export interface DriveFile {
  id: string;
  kind: string;
  mimeType: string;
  name: string;
}

/**
 * Models the file metadata
 * to be sent to Google Drive
 */
export interface DriveFileMetadata {
  name: string;
  parents?: string[];
}

/**
 * Base paths that we need
 * to interact with Google Drive
 */
export enum DrivePaths {
  API = 'https://www.googleapis.com',
  OAUTH2 = 'https://accounts.google.com/o/oauth2/v2/auth',
}
