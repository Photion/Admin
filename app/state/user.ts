
import { reactive } from '@vue/composition-api';


export const user = reactive({
  name: '',
  password: '',
});
