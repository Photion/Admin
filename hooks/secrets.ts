import { ref } from '@vue/composition-api';
import { handler, creds } from '@/hooks/encryption';


export enum Secrets {
  AWS_ACCESS_KEY_ID = 'AWS_ACCESS_KEY_ID',
  AWS_REGION = 'AWS_REGION',
  AWS_SECRET_ACCESS_KEY = 'AWS_SECRET_ACCESS_KEY',
  GCP_API_KEY = 'GCP_API_KEY',
  GCP_API_SECRET = 'GCP_API_SECRET',
}

export const PHOTION_SESSION_CREDENTIALS = 'PHOTION_SESSION_CREDENTIALS';

export const secrets = {
  values: {
    AWS_ACCESS_KEY_ID: ref(''),
    AWS_REGION: ref('eu-west-1'),
    AWS_SECRET_ACCESS_KEY: ref(''),
    GCP_API_KEY: ref(''),
    GCP_API_SECRET: ref(''),
  },

  entries: () => Object.entries(secrets.values),

  list: () => Object.values(secrets.values),

  serialize: () => secrets.entries().map(([key, obj]) => [key, obj.value]),

  deserialize: (payload: { secrets: string[][] }) => {
    payload.secrets.forEach(([key, value]) => {
      const secret = secrets.values[key as Secrets];

      if (secret) {
        secret.value = value;
        return true;
      }

      return false;
    });
  },

  save: async () => {
    const serialized = secrets.serialize();
    window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(serialized));

    await handler.save({ password: creds.value }, { secrets: secrets.serialize() });
  },

  load: async () => {
    const payload = await handler.load({ password: creds.value }) as { secrets: string[][] };
    window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(payload.secrets));
    secrets.deserialize(payload);
  },

  init: () => {
    const encoded = window.sessionStorage.getItem(PHOTION_SESSION_CREDENTIALS);

    if (encoded) {
      return secrets.deserialize({ secrets: JSON.parse(encoded) });
    }

    return false;
  },
};
