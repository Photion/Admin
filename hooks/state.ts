import { ref, Ref } from '@vue/composition-api';

import { Concept } from '@/app/models/Concept';
import { Media } from '@/app/models/Media';
import { Project } from '@/app/models/Project';

export const media: Ref<{
  concepts: { [key: string]: Concept },
  media: { [key: string]: Media },
  projects: { [key: string]: Project },
}> = ref({
  concepts: {},
  media: {},
  projects: {},
});

export const site = ref({
  domain: '',
  title: '',
  description: '',
});

export const user = ref({
  name: '',
  email: '',
  bio: '',
});

export const useState = () => ({
  media,
  site,
  user,
});
