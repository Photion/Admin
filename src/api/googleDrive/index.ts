export * from '~/src/api/googleDrive/api';
export * from '~/src/api/googleDrive/client';
export * from '~/src/api/googleDrive/types';
