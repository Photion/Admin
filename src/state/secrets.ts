import { computed, reactive } from '@vue/composition-api';
import { LocalStorageHandler } from 'keylocal/storage/localStorage';
import { WebCryptoAesGcpHandler } from 'keylocal/strategy/web/aes-gcp';

import { user } from '~/src/state/user';


export const PHOTION_ENCRYPTION_STORAGE_KEY = 'PHOTION_ENCRYPTION_STORAGE_KEY';
export const PHOTION_SESSION_CREDENTIALS = 'PHOTION_SESSION_CREDENTIALS';

export const strategy = new WebCryptoAesGcpHandler();

export const handler = new LocalStorageHandler(PHOTION_ENCRYPTION_STORAGE_KEY, {
  strategy,
  localStorage: window.localStorage,
});

export const encryptionKey = computed(() => `${user.name}#${user.password}`);

export const encrypt = (plaintext: string): Promise<string> => strategy.encrypt({ password: encryptionKey.value }, plaintext);

export const decrypt = (encrypted: string): Promise<string> => strategy.decrypt({ password: encryptionKey.value }, encrypted);

export const secrets = reactive({
  aws: {
    region: '',
    accessKeyId: '',
    secretAccessKey: '',
  },
  gcp: {
    apiKey: '',
    apiSecret: '',
  },
});

export const deserialize = (payload: any) => {
  secrets.aws.region = payload?.aws?.region || '';
  secrets.aws.accessKeyId = payload?.aws?.accessKeyId || '';
  secrets.aws.secretAccessKey = payload?.aws?.secretAccessKey || '';
  secrets.gcp.apiKey = payload?.gcp?.apiKey || '';
  secrets.gcp.apiSecret = payload?.gcp?.apiSecret || '';
};

export const save = async () => {
  window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(secrets));

  await handler.save({ password: encryptionKey.value }, secrets);
};

export const load = async () => {
  const secrets = await handler.load({ password: encryptionKey.value });

  window.sessionStorage.setItem(PHOTION_SESSION_CREDENTIALS, JSON.stringify(secrets));

  deserialize(secrets);
};

export const init = () => {
  const encoded = window.sessionStorage.getItem(PHOTION_SESSION_CREDENTIALS);

  if (encoded) {
    deserialize(JSON.parse(encoded));
  }

  user.name = window.sessionStorage.getItem('PHOTION_USERNAME') || '';
};
