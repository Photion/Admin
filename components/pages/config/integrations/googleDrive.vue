<template>
  <div>
    <h3>Integrate with Google Drive</h3>
    <div>
      Not supported yet.
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';
import { secrets } from '@/app/state/secrets';

export default defineComponent({

  setup() {
    return {
      secrets,
    };
  },
});
</script>
