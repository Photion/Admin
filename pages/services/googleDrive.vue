<template>
  <service-form :service="googleDrive" preview>
    <div>
      No fields.
    </div>
  </service-form>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';

import { secrets } from '@/app/state/secrets';
import { googleDrive } from '@/app/state/service';
import ServiceForm from '@/components/pages/services/ServiceForm.vue';

export default defineComponent({

  components: {
    ServiceForm,
  },

  setup() {
    return {
      googleDrive,
      secrets,
    };
  },
});
</script>
