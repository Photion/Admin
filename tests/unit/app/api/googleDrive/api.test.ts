/* eslint-disable @typescript-eslint/camelcase */
import axios from 'axios';

import { DriveScopes } from '~/src/api/googleDrive/types';
import { secrets } from '~/src/state/secrets';
import {
  getOAuthUrl,
  readHashSearchString,
  createFile,
  createFolder,
  updateFile,
  downloadFile,
  deleteFile,
  searchFiles,
} from '~/src/api/googleDrive/api';

describe('unit.app.api.googleDrive.api', () => {
  const client_id = '12345';
  const redirect_uri = 'https://admin.photion.app';

  beforeAll(() => {
    secrets.googleDrive.accessToken = 'a';
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('getOAuthUrl', () => {
    [
      { scope: '' },
      { scope: DriveScopes.DRIVE__APPDATA },
      { scope: DriveScopes.DRIVE__FILE },
      { scope: DriveScopes.DRIVE__METADATA },
      { scope: Object.values(DriveScopes).join(' ') },
    ]
      .forEach(({ scope }) => {
        it('Generates an auth url with the given scope', () => {
          const actual = getOAuthUrl({ redirect_uri, client_id, scope });
          expect(actual).toMatchSnapshot();
        });
      });
  });

  describe('readHashSearchString', () => {
    [
      { url: redirect_uri, expected: {} },
      { url: redirect_uri + '?param1=1', expected: {} },
      { url: redirect_uri + '#param2=2', expected: { param2: '2' } },
      { url: redirect_uri + '?param1=1#param2=2', expected: { param2: '2' } },
      { url: redirect_uri + '?param1=1&paramq=q', expected: {} },
      { url: redirect_uri + '?param1=1&paramq=q#param2=2', expected: { param2: '2' } },
      { url: redirect_uri + '?param1=1&paramq=q#param2=2&paramt=t', expected: { param2: '2', paramt: 't' } },
      { url: redirect_uri + '#param2=1&param2=2', expected: { param2: '2' } },
    ]
      .forEach((testCase) => {
        it('Parses the hash string as it was a query string', () => {
          const actual = readHashSearchString(testCase.url);
          expect(actual).toEqual(testCase.expected);
        });
      });
  });

  describe('createFile', () => {
    it('Queries with axios', async () => {
      const expected = 'createFile';
      const metadata = {
        name: 'file',
      };

      const file = new File([], 'file');

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await createFile(metadata, file);
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('createFolder', () => {
    it('Queries with axios', async () => {
      const expected = 'createFolder';

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await createFolder('my-folder');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('updateFile', () => {
    it('Queries with axios without a file', async () => {
      const expected = 'updateFile';
      const metadata = {
        name: 'file',
      };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await updateFile('my-file-id', metadata);
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios with a file', async () => {
      const expected = 'updateFile';
      const metadata = {
        name: 'file',
      };
      const file = new File([], 'file');

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await updateFile('my-file-id', metadata, file);
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('downloadFile', () => {
    it('Queries with axios asking for a blob', async () => {
      const expected = { data: 'downloadFile' };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await downloadFile('my-file', 'blob');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios asking for a json', async () => {
      const expected = { data: 'downloadFile' };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await downloadFile('my-file', 'json');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('deleteFile', () => {
    it('Queries with axios', async () => {
      const expected = 'deleteFile';

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await deleteFile('my-folder');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('searchFiles', () => {
    it('Queries with axios for all files', async () => {
      const expected = { data: { files: [1, 2, 3] } };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await searchFiles();
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data.files);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios for all files with a name', async () => {
      const expected = { data: { files: [1, 2, 3] } };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await searchFiles('my-file-name');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data.files);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios for all files in a parent', async () => {
      const expected = { data: { files: [1, 2, 3] } };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await searchFiles(undefined, 'my-parent');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data.files);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios for all files with a name in a parent', async () => {
      const expected = { data: { files: [1, 2, 3] } };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await searchFiles('my-file-name', 'my-parent');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data.files);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

    it('Queries with axios for all files and their fields', async () => {
      const expected = { data: { files: [1, 2, 3] } };

      jest.spyOn(axios, 'request').mockImplementationOnce(() => Promise.resolve(expected));

      const actual = await searchFiles(undefined, undefined, '*');
      const mock = (axios.request as unknown as { mock: jest.MockContext<typeof axios.request, unknown[]> }).mock;

      expect(actual).toEqual(expected.data.files);
      expect(mock.calls[0][0]).toMatchSnapshot();
    });

  });


});
