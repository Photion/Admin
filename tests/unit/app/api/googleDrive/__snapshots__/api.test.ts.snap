// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.app.api.googleDrive.api createFile Queries with axios 1`] = `
Object {
  "data": FormData {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "POST",
  "url": "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart",
}
`;

exports[`unit.app.api.googleDrive.api createFolder Queries with axios 1`] = `
Object {
  "data": FormData {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "POST",
  "url": "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart",
}
`;

exports[`unit.app.api.googleDrive.api deleteFile Queries with axios 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "DELETE",
  "url": "https://www.googleapis.com/drive/v2/files/my-folder",
}
`;

exports[`unit.app.api.googleDrive.api downloadFile Queries with axios asking for a blob 1`] = `
Object {
  "data": Object {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "responseType": "blob",
  "url": "https://www.googleapis.com/drive/v2/files/my-file?alt=media",
}
`;

exports[`unit.app.api.googleDrive.api downloadFile Queries with axios asking for a json 1`] = `
Object {
  "data": Object {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "responseType": "json",
  "url": "https://www.googleapis.com/drive/v2/files/my-file?alt=media",
}
`;

exports[`unit.app.api.googleDrive.api getOAuthUrl Generates an auth url with the given scope 1`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=token&client_id=12345&redirect_uri=https%3A%2F%2Fadmin.photion.app&scope=&state="`;

exports[`unit.app.api.googleDrive.api getOAuthUrl Generates an auth url with the given scope 2`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=token&client_id=12345&redirect_uri=https%3A%2F%2Fadmin.photion.app&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.appdata&state="`;

exports[`unit.app.api.googleDrive.api getOAuthUrl Generates an auth url with the given scope 3`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=token&client_id=12345&redirect_uri=https%3A%2F%2Fadmin.photion.app&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.file&state="`;

exports[`unit.app.api.googleDrive.api getOAuthUrl Generates an auth url with the given scope 4`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=token&client_id=12345&redirect_uri=https%3A%2F%2Fadmin.photion.app&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.metadata&state="`;

exports[`unit.app.api.googleDrive.api getOAuthUrl Generates an auth url with the given scope 5`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=token&client_id=12345&redirect_uri=https%3A%2F%2Fadmin.photion.app&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.appdata%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.file%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.metadata&state="`;

exports[`unit.app.api.googleDrive.api searchFiles Queries with axios for all files 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "params": Object {
    "q": "",
  },
  "url": "https://www.googleapis.com/drive/v3/files",
}
`;

exports[`unit.app.api.googleDrive.api searchFiles Queries with axios for all files and their fields 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "params": Object {
    "fields": "*",
    "q": "",
  },
  "url": "https://www.googleapis.com/drive/v3/files",
}
`;

exports[`unit.app.api.googleDrive.api searchFiles Queries with axios for all files in a parent 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "params": Object {
    "q": "'my-parent' in parents",
  },
  "url": "https://www.googleapis.com/drive/v3/files",
}
`;

exports[`unit.app.api.googleDrive.api searchFiles Queries with axios for all files with a name 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "params": Object {
    "q": "name='my-file-name'",
  },
  "url": "https://www.googleapis.com/drive/v3/files",
}
`;

exports[`unit.app.api.googleDrive.api searchFiles Queries with axios for all files with a name in a parent 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "GET",
  "params": Object {
    "q": "name='my-file-name' and 'my-parent' in parents",
  },
  "url": "https://www.googleapis.com/drive/v3/files",
}
`;

exports[`unit.app.api.googleDrive.api updateFile Queries with axios with a file 1`] = `
Object {
  "data": FormData {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "PUT",
  "url": "https://www.googleapis.com/upload/drive/v2/files/my-file-id?uploadType=multipart",
}
`;

exports[`unit.app.api.googleDrive.api updateFile Queries with axios without a file 1`] = `
Object {
  "data": FormData {},
  "headers": Object {
    "Authorization": "Bearer a",
  },
  "method": "PUT",
  "url": "https://www.googleapis.com/drive/v2/files/my-file-id?uploadType=multipart",
}
`;
